module/zfs/abd.c

```diff
@@ -247,9 +346,11 @@ static inline abd_t *
 abd_alloc_struct(size_t chunkcnt)
 {
 	size_t size = offsetof(abd_t, abd_u.abd_scatter.abd_chunks[chunkcnt]);
-	abd_t *abd = kmem_alloc(size, KM_PUSHPAGE);
+	abd_t *abd = kmem_zalloc(size, KM_PUSHPAGE);
 	ASSERT3P(abd, !=, NULL);
 	ABDSTAT_INCR(abdstat_struct_size, size);
+	abd->abd_create_time = gethrtime();
+	mutex_init(&abd->abd_mutex, NULL, MUTEX_DEFAULT, NULL);
```

Comment:
Yes, abd_struct_alloc() is called each time a new abd is created (e.g. whenever a buffer is read into memory or is dirtied).

abd_struct_alloc() serves out the memory structure from the heap which is heavily kmem cached.   Unfortunately the function can't easily have its own kmem cache because the structures are of variable size; you'd have to set up a bunch of different kmem caches as is done in zio.c (and for the kmem caches themselves) and that doesn't seem worth it.

I'm not sure what kmem_cache has to do with a mutex_init().   The mutex is part of the header structure, as it protects the individual header structure itself, and mutex_init() is (when not debugging mutexes) pretty cheap.